File: src/test/java/no/novari/flyt/integration/validation/constraints/ReferencedConfigurationIsForIntegrationValidatorTest.java
package no.novari.flyt.integration.validation.constraints;

import no.novari.flyt.integration.model.dtos.ConfigurationDto;
import no.novari.flyt.integration.validation.IntegrationValidationContext;
import org.hibernate.validator.constraintvalidation.HibernateConstraintValidatorContext;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class ReferencedConfigurationIsForIntegrationValidatorTest {

    private ReferencedConfigurationIsForIntegrationValidator validator;
    private IntegrationValidationContext validationContext;

    @BeforeEach
    void setUp() {
        validator = new ReferencedConfigurationIsForIntegrationValidator();
        validationContext = mock(IntegrationValidationContext.class);
    }

    @Test
    void testIsValid_NullValue_ReturnsTrue() {
        boolean result = validator.isValid(null, createMockContext());

        assertTrue(result);
    }

    @Test
    void testIsValid_SameIntegrationId_ReturnsTrue() {
        long integrationId = 12345L;
        when(validationContext.getIntegrationId()).thenReturn(integrationId);
        when(validationContext.getConfiguration()).thenReturn(
                ConfigurationDto.builder()
                        .integrationId(integrationId)
                        .build()
        );
        HibernateConstraintValidatorContext mockContext = createMockContext();

        boolean result = validator.isValid(1L, mockContext);

        assertTrue(result);
    }

    @Test
    void testIsValid_DifferentIntegrationId_ReturnsFalse() {
        long integrationId = 12345L;
        long differentIntegrationId = 67890L;
        when(validationContext.getIntegrationId()).thenReturn(integrationId);
        when(validationContext.getConfiguration()).thenReturn(
                ConfigurationDto.builder()
                        .integrationId(differentIntegrationId)
                        .build()
        );
        HibernateConstraintValidatorContext mockContext = createMockContext();

        boolean result = validator.isValid(1L, mockContext);

        assertFalse(result);
    }

    private HibernateConstraintValidatorContext createMockContext() {
        HibernateConstraintValidatorContext mockContext = mock(HibernateConstraintValidatorContext.class);
        when(mockContext.getConstraintValidatorPayload(IntegrationValidationContext.class))
                .thenReturn(validationContext);
        return mockContext;
    }

}

File: src/test/kotlin/no/novari/flyt/integration/validation/constraints/ReferencedConfigurationIsForIntegrationValidatorTest.kt
package no.novari.flyt.integration.validation.constraints

import no.novari.flyt.integration.api.dto.ConfigurationDto
import no.novari.flyt.integration.validation.IntegrationValidationContext
import org.hibernate.validator.constraintvalidation.HibernateConstraintValidatorContext
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

class ReferencedConfigurationIsForIntegrationValidatorTest {
    private lateinit var validator: ReferencedConfigurationIsForIntegrationValidator
    private lateinit var validationContext: IntegrationValidationContext

    @BeforeEach
    fun setUp() {
        validator = ReferencedConfigurationIsForIntegrationValidator()
        validationContext = mock()
    }

    @Test
    fun shouldReturnTrueForNullValue() {
        val result = validator.isValid(null, createMockContext())

        assertTrue(result)
    }

    @Test
    fun shouldReturnTrueWhenConfigurationBelongsToIntegration() {
        whenever(validationContext.integrationId).thenReturn(12345L)
        whenever(validationContext.configuration).thenReturn(ConfigurationDto(integrationId = 12345L))

        val result = validator.isValid(1L, createMockContext())

        assertTrue(result)
    }

    @Test
    fun shouldReturnFalseWhenConfigurationBelongsToAnotherIntegration() {
        whenever(validationContext.integrationId).thenReturn(12345L)
        whenever(validationContext.configuration).thenReturn(ConfigurationDto(integrationId = 67890L))

        val result = validator.isValid(1L, createMockContext())

        assertFalse(result)
    }

    private fun createMockContext(): HibernateConstraintValidatorContext {
        val mockContext = mock<HibernateConstraintValidatorContext>()
        whenever(mockContext.getConstraintValidatorPayload(IntegrationValidationContext::class.java))
            .thenReturn(validationContext)
        return mockContext
    }
}
